### ISSUE TITLE
Discuss adding support for Decentrelised Exchange 

### ISSUE THREAD

**Issue by captainreptile** [ctubio/Krypto-trading-bot#578]
Greetings,

We all love our decentralized cryptos _(not all cryptos are)_, but still risk them by sending to a centralized exchanges.  To a one man show and one man hands - Who can anytime decide to stop withdrawals or close the bazar.

It could be pretty amazing if we add support for Decentralized exchanges
https://openledger.io
https://bisq.network
https://beta.wavesplatform.com/create
https://crypto-bridge.org/

Please share your general and technical thoughts on this

**Comment by woshikie** [comment]
Would love to, but the API implementations are all private, i believe

**Comment by ctubio** [comment]
yep the private key is in my hands; will consider soOn the addition of each of these exchanges


many thanks for the suggestion'''''

**Comment by captainreptile** [comment]
From the list above i believe **Bisq** is the only 100% Decentralized as they combine the coolest tech: P2P Over Tor network,No server to DDOS And a wikipedia like Human Arbitrators. _(Torrent left behind)_

I was only afraid that there is no sort of API.. but there is some good news:

https://github.com/mrosseel/bisq-proposals/blob/api-proposal/http-api.adoc
https://mrosseel.github.io/bisq-api-examples/
https://github.com/mrosseel/bisq-api
https://bisq.community/t/announcing-the-trading-api/3424

As i understand Local Bisq application acts as server for the api.

Possible Downsides:
- Tor can be slow
- Bot might end up trading with only other trading bots or maybe non:
Some orders require human action (like send money from bank or wallet). its up for the user to extend the api to automate the trade (like integrating bank api or ethereum wallet api)

**Comment by donaldforest** [comment]
Is always cool when we have more choice to using more exchanges. For me is most necessary rules and factors to add new exchange:
- low fees, and again LOW FEES!
- fucking stable and adequate API (no comments)
- trade volumes

All another isn't matter.
Decentralised Exchange - yes it cool also.

**Comment by gabev** [comment]
Bitshare , Openledger, crypto-bridge  they use Graphene API ...the documentation for the API is found here : http://docs.bitshares.org/api/index.html ...as a matter of fact there is a working proposal for the creation of a simple trading bot with simple strategies to be integrated into the Dex.. the telegram group where those issues are discussed is here : https://t.me/BTSLiquidityBotWorkerProposal ...would be awesome if we can integrate the Dex into Krypto.

**Comment by donaldforest** [comment]
https://www.altcoin.io/

**Comment by tiimbo82** [comment]
idex.market

**Comment by bdbabiak** [comment]
cryptobridge - yes!

bitshares api, right?

**Comment by ThVWh** [comment]
DDEX now available with the Hydro API.

Also, listed #2 in volume on ERC20 tokens (behind Bancor) as per Etherscan.io
https://etherscan.io/stat/dextracker

**Comment by xloem** [comment]
I was looking into https://stellar.org/, which appeared larger and more versatile to me than a couple others I briefly reviewed.

Since this task is both large and important, I think if we started making progress on this, we could work together to accomplish it, if people just contributed a little bit when they were able to.  We could perhaps wrap the proprietary binary library with an open source one to accomplish this.

It would be helpful to agree on a DEX to start with.  I guess it should be truly decentralized and provide for rapid order placement and streamed orderbook updates.  Can anybody verify that one of the DEX's meets this requirement?  Can anybody think of any other requirements?

**Comment by ctubio** [comment]
let me see the apis of all suggested and i will try to implement one that have websockets; is long time since a new api is not added :D

**Comment by gabev** [comment]
Crypto-Bridge, OpenLedger, EasyDEx  and the rest for a total of 16 decentralized exchanges so far, are all using the BitShares API. Documentation here: BitShares Developer Portal- https://dev.bitshares.works/en/master/ . In term of speed , 3k+ tps, block time 1-3 sec, atomic swaps on chain..regarding DEXBot , is a pure market making ..as of right now we are in the process of refactoring the code for implementing async and speed up the bot...hope that helps

**Comment by xloem** [comment]
hey @gabev, I was looking at bitshares, and I noticed there is a flat transaction fee for every order placed and every order canceled.  This worries me a bit, because Krypto behaves such that it places and cancels orders very frequently.  I'm worried the transaction fees could add up.  Do you have any thoughts on that?

**Comment by MarkoPaasila** [comment]
@xloem Yes on Bitshares there's both a flat fee for placing an order and a trade fee. The trade fee depends on the asset, and is usually between 0% and 0.2%. The order placement fee is relatively small, and is returned if the order is cancelled. In my experience, $1 will keep me going for a year per strategy. We take the order fee's into account in DEXBot when calculating order placement - to prevent negative returns.

There's a good Python library for Bitshares (pip install bitshares), with key/wallet management. You can see how we use it in DEXBot. We are in the process of separating the code into separate engines for getting data and for placing orders. It's open source (MIT), so you can copy all you like.

The thing with any DEX is that ticks come in slowly. If you want high frequency trading, you probably need to integrate a P2P node into it in order to get pre-consensus data. On Bitshares ticks are 3s, on EOS 0.5s, and Ethereum based ones something like a minute (not sure).

**Comment by xloem** [comment]
@MarkoPaasila thanks for the information that placement fees are refunded on cancel.

It looks like bitshares allows for high-frequency information via the [set_pending_transaction_callback](https://dev.bitshares.works/en/master/api/websocket.html#database-notifications) api call.  This seems to send order notifications on the websocket as soon as they come in, before the block is confirmed.
